File: client-android/lib/src/main/kotlin/stasis/client_android/lib/utils/Cache.kt
package stasis.client_android.lib.utils

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import stasis.client_android.lib.utils.NonFatal.nonFatal
import stasis.client_android.lib.utils.Try.Companion.flatMap
import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference
import kotlin.io.path.deleteIfExists
import kotlin.io.path.readBytes
import kotlin.io.path.writeBytes
import kotlin.math.min
import kotlin.streams.toList as kotlinToList

interface Cache<K : Any, V> {

    /**
     * Retrieves the cached value associated with [key], if any.
     *
     * @param key the key associated with the needed value
     * @return the cached value or `null` if none was found
     */
    suspend fun get(key: K): V?

    /**
     * Puts a new [value] with the provided [key] in the cache.
     *
     * @param key the key associated with the value
     * @param value the value to insert
     */
    suspend fun put(key: K, value: V)

    /**
     * Puts the provided entries in the cache.
     *
     * @param entries entries to insert
     */
    suspend fun put(entries: kotlin.collections.Map<K, V>)

    /**
     * Retrieves the cached value associated with [key] or attempts to load it.
     *
     * @param key the key associated with the needed value
     * @param load function used for loading the value if it is not already in the cache
     * @return the cached or loaded value
     */
    suspend fun getOrLoad(key: K, load: suspend (K) -> V?): V?

    /**
     * Removes the cached value associated with [key], if any.
     *
     * @param key the key associated with value to be removed
     */
    suspend fun remove(key: K)

    /**
     * Retrieves all cached values.
     */
    suspend fun all(): kotlin.collections.Map<K, V>

    /**
     * Removes all cached values.
     */
    suspend fun clear()

    /**
     * Provides the current read statistics for this cache.
     */
    val readStatistics: OperationStatistics

    /**
     * Provides the current write statistics for this cache.
     */
    val writeStatistics: OperationStatistics

    /**
     * Simple in-memory cache backed by [ConcurrentHashMap].
     */
    class Map<K : Any, V> : Cache<K, V> {
        private val map: ConcurrentHashMap<K, V> = ConcurrentHashMap()

        override suspend fun get(key: K): V? = map[key]

        override suspend fun put(key: K, value: V) {
            map[key] = value
        }

        override suspend fun put(entries: kotlin.collections.Map<K, V>) =
            map.putAll(entries)

        override suspend fun getOrLoad(key: K, load: suspend (K) -> V?): V? =
            get(key) ?: load(key)?.also { map[key] = it }

        override suspend fun remove(key: K) {
            map.remove(key)
        }

        override suspend fun all(): kotlin.collections.Map<K, V> =
            map

        override suspend fun clear() =
            map.clear()

        override val readStatistics: OperationStatistics = OperationStatistics.empty()

        override val writeStatistics: OperationStatistics = OperationStatistics.empty()
    }

    /**
     * File-based cache.
     *
     * @param target directory for storing state files
     * @param serdes key and value de/serializers
     */
    class File<K : Any, V>(
        private val target: Path,
        private val serdes: Serdes<K, V>
    ) : Cache<K, V> {
        private val stats: IoStatistics = IoStatistics()

        private val inMemory = Map<K, V>()

        override suspend fun get(key: K): V? = inMemory.get(key) ?: Try {
            withContext(Dispatchers.IO) {
                val start = System.currentTimeMillis()
                val bytes = key.asStateFile().readBytes()
                val duration = System.currentTimeMillis() - start
                bytes to duration
            }
        }.flatMap { (bytes, duration) ->
            stats.bytesRead(amount = bytes.size, duration = duration)
            serdes.deserializeValue(bytes)
        }.toOption()?.also { inMemory.put(key, it) }

        override suspend fun put(key: K, value: V) {
            inMemory.put(key, value)
            withContext(Dispatchers.IO) {
                val start = System.currentTimeMillis()
                val bytes = serdes.serializeValue(value)
                key.asStateFile().writeBytes(bytes)
                val duration = System.currentTimeMillis() - start
                stats.bytesWritten(amount = bytes.size, duration = duration)
            }
        }

        override suspend fun put(entries: kotlin.collections.Map<K, V>) =
            entries.forEach { put(it.key, it.value) }

        override suspend fun getOrLoad(key: K, load: suspend (K) -> V?): V? =
            get(key) ?: load(key)?.also { put(key, it) }

        override suspend fun remove(key: K) {
            inMemory.remove(key)
            withContext(Dispatchers.IO) {
                key.asStateFile().deleteIfExists()
            }
        }

        override suspend fun all(): kotlin.collections.Map<K, V> =
            keys()
                .mapNotNull { key -> get(key)?.let { value: V -> key to value } }
                .toMap()

        override suspend fun clear() =
            keys().forEach { remove(it) }

        override val readStatistics: OperationStatistics
            get() = stats.readStatistics

        override val writeStatistics: OperationStatistics
            get() = stats.writeStatistics

        private fun K.asStateFile(): Path {
            Files.createDirectories(target)
            return target.resolve(serdes.serializeKey(this))
        }

        private fun keys(): List<K> =
            Try { Files.list(target).kotlinToList() }
                .getOrElse { emptyList() }
                .mapNotNull { serdes.deserializeKey(it.fileName.toString()) }

        interface Serdes<K : Any, V> {
            fun serializeKey(key: K): String
            fun deserializeKey(key: String): K?
            fun serializeValue(value: V): ByteArray
            fun deserializeValue(value: ByteArray): Try<V>
        }
    }

    /**
     * Time-based expiration cache.
     *
     * Newly loaded values are scheduled to expire after the specified amount of time has passed.
     *
     * @param underlying cache that handles the actual data storage
     * @param expiration amount of time to keep values in the cache
     * @param scope coroutine scope for executing expiration tasks
     */
    class Expiring<K : Any, V>(
        private val underlying: Cache<K, V>,
        private val expiration: Duration,
        private val scope: CoroutineScope,
    ) : Cache<K, V> {
        private val listeners: ConcurrentLinkedQueue<(K) -> Unit> = ConcurrentLinkedQueue()

        private val jobs: ConcurrentHashMap<K, Job> = ConcurrentHashMap()

        override suspend fun get(key: K): V? =
            underlying.get(key)

        override suspend fun put(key: K, value: V) {
            underlying.put(key, value)
            expire(key)
        }

        override suspend fun put(entries: kotlin.collections.Map<K, V>) {
            entries.forEach { put(it.key, it.value) }
        }

        override suspend fun getOrLoad(key: K, load: suspend (K) -> V?): V? =
            underlying.getOrLoad(key, load = { k -> load(k)?.also { expire(k) } })

        override suspend fun remove(key: K) {
            jobs.remove(key)?.cancel()
            underlying.remove(key)
        }

        override suspend fun all(): kotlin.collections.Map<K, V> =
            underlying.all()

        override suspend fun clear() {
            jobs.forEach { it.value.cancel() }
            jobs.clear()
            underlying.clear()
        }

        override val readStatistics: OperationStatistics
            get() = underlying.readStatistics

        override val writeStatistics: OperationStatistics
            get() = underlying.writeStatistics

        fun registerOnEntryExpiredListener(listener: (K) -> Unit) {
            listeners.add(listener)
        }

        fun unregisterOnEntryExpiredListener(listener: (K) -> Unit) {
            listeners.remove(listener)
        }

        private fun expire(key: K) {
            try {
                val job = scope.launch {
                    delay(timeMillis = expiration.toMillis())
                    remove(key)
                    listeners.forEach { it(key) }
                }

                jobs[key] = job
            } catch (_: CancellationException) {
                // do nothing
            }
        }
    }

    /**
     * Time-based refreshing cache.
     *
     * After the specified amount of time has passed, values are reloaded using the `load` function
     * initially associated with a key. This means that using different `load` functions for subsequent
     * calls to [getOrLoad] will not be taken into account during refreshes.
     *
     * Care needs to be taken to avoid multiple `load` functions as this will lead to inconsistencies
     * when the [underlying] cache loads values and this cache refreshes them. The proper way to reset
     * a `load` function is to [remove] the entry and re-add it with the new function.
     *
     * @param underlying cache that handles the actual data storage
     * @param interval amount of time to wait before refreshing values
     * @param scope coroutine scope for executing refresh tasks
     */
    class Refreshing<K : Any, V>(
        private val underlying: Cache<K, V>,
        private val interval: Duration,
        private val scope: CoroutineScope
    ) : Cache<K, V> {
        private val listeners: ConcurrentLinkedQueue<(K, V?) -> Unit> = ConcurrentLinkedQueue()

        private val jobs: ConcurrentHashMap<K, Job> = ConcurrentHashMap()

        override suspend fun get(key: K): V? =
            underlying.get(key)

        override suspend fun put(key: K, value: V) {
            underlying.put(key, value)
        }

        override suspend fun put(entries: kotlin.collections.Map<K, V>) {
            entries.forEach { put(it.key, it.value) }
        }

        override suspend fun getOrLoad(key: K, load: suspend (K) -> V?): V? =
            underlying.getOrLoad(key, load)?.also {
                if (!jobs.containsKey(key)) {
                    refresh(key, load, interval)
                }
            }

        override suspend fun remove(key: K) {
            jobs.remove(key)?.cancel()
            underlying.remove(key)
        }

        override suspend fun all(): kotlin.collections.Map<K, V> =
            underlying.all()

        override suspend fun clear() {
            jobs.forEach { it.value.cancel() }
            jobs.clear()
            underlying.clear()
        }

        override val readStatistics: OperationStatistics
            get() = underlying.readStatistics

        override val writeStatistics: OperationStatistics
            get() = underlying.writeStatistics

        fun registerOnEntryRefreshedListener(listener: (K, V?) -> Unit) {
            listeners.add(listener)
        }

        fun unregisterOnEntryRefreshedListener(listener: (K, V?) -> Unit) {
            listeners.remove(listener)
        }

        private fun refresh(key: K, load: suspend (K) -> V?, refreshInterval: Duration) {
            try {
                val job = scope.launch {
                    delay(timeMillis = refreshInterval.toMillis())

                    try {
                        val refreshedValue = load(key)

                        when (refreshedValue) {
                            null -> underlying.remove(key)
                            else -> underlying.put(key, refreshedValue)
                        }

                        listeners.forEach { it(key, refreshedValue) }
                        refresh(key, load, interval)
                    } catch (e: Throwable) {
                        e.nonFatal()
                        val nextRefresh = Duration.ofMillis(
                            min(
                                interval.dividedBy(RefreshIntervalOnFailureReduction).toMillis(),
                                MaxRefreshIntervalOnFailure.toMillis()
                            )
                        )
                        refresh(key, load, nextRefresh)
                    }
                }

                jobs[key] = job
            } catch (_: CancellationException) {
                // do nothing
            }
        }

        companion object {
            const val RefreshIntervalOnFailureReduction: Long = 10L
            val MaxRefreshIntervalOnFailure: Duration = Duration.ofSeconds(5)
        }
    }

    /**
     * Statistics-tracking cache.
     *
     * Statistics about cache hits and misses are kept and provided.
     *
     * @param underlying cache that handles the actual data storage
     */
    class Tracking<K : Any, V>(
        private val underlying: Cache<K, V>,
    ) : Cache<K, V> {
        private val stats: HitStatistics = HitStatistics()

        /**
         * Provides the current number of cache hits.
         */
        val hits: Long
            get() = stats.hits

        /**
         * Provides the current number of cache misses.
         */
        val misses: Long
            get() = stats.misses

        override val readStatistics: OperationStatistics
            get() = underlying.readStatistics

        override val writeStatistics: OperationStatistics
            get() = underlying.writeStatistics

        override suspend fun get(key: K): V? {
            val result = underlying.get(key = key)

            when (result) {
                null -> stats.miss()
                else -> stats.hit()
            }

            return result
        }

        override suspend fun put(key: K, value: V) =
            underlying.put(key = key, value = value)

        override suspend fun put(entries: kotlin.collections.Map<K, V>) =
            underlying.put(entries = entries)

        override suspend fun getOrLoad(key: K, load: suspend (K) -> V?): V? =
            get(key) ?: load(key)?.also { put(key, it) }

        override suspend fun remove(key: K) =
            underlying.remove(key = key)

        override suspend fun all(): kotlin.collections.Map<K, V> =
            underlying.all()

        override suspend fun clear() =
            underlying.clear()
    }

    private class HitStatistics {
        private val hitsRef: AtomicLong = AtomicLong(0)
        private val missesRef: AtomicLong = AtomicLong(0)

        fun hit() {
            hitsRef.incrementAndGet()
        }

        fun miss() {
            missesRef.incrementAndGet()
        }

        val hits: Long
            get() = hitsRef.get()

        val misses: Long
            get() = missesRef.get()
    }

    private class IoStatistics {
        private val readStatsRef: AtomicReference<OperationStatistics> =
            AtomicReference(OperationStatistics.empty())

        private val writeStatsRef: AtomicReference<OperationStatistics> =
            AtomicReference(OperationStatistics.empty())

        fun bytesRead(amount: Int, duration: Long) {
            readStatsRef.accumulateAndGet(null) { stats, _ -> stats.updateWith(amount, duration) }
        }

        fun bytesWritten(amount: Int, duration: Long) {
            writeStatsRef.accumulateAndGet(null) { stats, _ -> stats.updateWith(amount, duration) }
        }

        val readStatistics: OperationStatistics
            get() = readStatsRef.get()

        val writeStatistics: OperationStatistics
            get() = writeStatsRef.get()
    }

    data class OperationStatistics(
        val bytesProcessed: Long,
        val minDuration: Long,
        val maxDuration: Long,
        val operations: Long,
    ) {
        fun updateWith(amount: Int, duration: Long): OperationStatistics =
            OperationStatistics(
                bytesProcessed = bytesProcessed + amount,
                minDuration = if (duration > minDuration) minDuration else duration,
                maxDuration = if (duration < maxDuration) maxDuration else duration,
                operations = operations + 1
            )

        companion object {
            fun empty(): OperationStatistics = OperationStatistics(
                bytesProcessed = 0,
                minDuration = Long.MAX_VALUE,
                maxDuration = Long.MIN_VALUE,
                operations = 0
            )
        }
    }
}

File: client-ios/lib/Sources/StasisClientLib/Utils/Cache.swift
import Foundation

public protocol Cache<Key, Value>: Sendable {
    associatedtype Key: Hashable & Sendable
    associatedtype Value: Sendable

    /// Retrieves the cached value associated with `key`, if any.
    ///
    /// - Parameter key: the key associated with the needed value
    /// - Returns: the cached value or `nil` if none was found
    func get(_ key: Key) async -> Value?

    /// Puts a new `value` with the provided `key` in the cache.
    ///
    /// - Parameters:
    ///   - key: the key associated with the value
    ///   - value: the value to insert
    func put(_ key: Key, _ value: Value) async throws

    /// Puts the provided entries in the cache.
    ///
    /// - Parameter entries: entries to insert
    func put(entries: [Key: Value]) async throws

    /// Retrieves the cached value associated with `key` or attempts to load it.
    ///
    /// - Parameters:
    ///   - key: the key associated with the needed value
    ///   - load: function used for loading the value if it is not already in the cache
    /// - Returns: the cached or loaded value
    func getOrLoad(_ key: Key, load: @Sendable @escaping (Key) async throws -> Value?) async throws -> Value?

    /// Removes the cached value associated with `key`, if any.
    ///
    /// - Parameter key: the key associated with value to be removed
    func remove(_ key: Key) async throws

    /// Retrieves all cached values.
    func all() async -> [Key: Value]

    /// Removes all cached values.
    func clear() async throws

    /// Provides the current read statistics for this cache.
    var readStatistics: OperationStatistics { get async }

    /// Provides the current write statistics for this cache.
    var writeStatistics: OperationStatistics { get async }
}

public struct OperationStatistics: Sendable, Equatable, Hashable {
    public let bytesProcessed: Int64
    public let minDuration: Int64
    public let maxDuration: Int64
    public let operations: Int64

    public init(bytesProcessed: Int64, minDuration: Int64, maxDuration: Int64, operations: Int64) {
        self.bytesProcessed = bytesProcessed
        self.minDuration = minDuration
        self.maxDuration = maxDuration
        self.operations = operations
    }

    public static func empty() -> OperationStatistics {
        OperationStatistics(
            bytesProcessed: 0,
            minDuration: .max,
            maxDuration: .min,
            operations: 0
        )
    }

    public func updateWith(amount: Int, duration: Int64) -> OperationStatistics {
        OperationStatistics(
            bytesProcessed: bytesProcessed + Int64(amount),
            minDuration: duration > minDuration ? minDuration : duration,
            maxDuration: duration < maxDuration ? maxDuration : duration,
            operations: operations + 1
        )
    }
}

public protocol CacheSerdes<Key, Value>: Sendable {
    associatedtype Key
    associatedtype Value

    func encodeKey(_ key: Key) -> String
    func decodeKey(_ key: String) -> Key?
    func encodeValue(_ value: Value) throws -> Data
    func decodeValue(_ data: Data) throws -> Value
}

/// Simple in-memory cache backed by a `Dictionary` inside an actor.
public actor MapCache<Key: Hashable & Sendable, Value: Sendable>: Cache {
    private var storage: [Key: Value] = [:]

    public init() {}

    public nonisolated var readStatistics: OperationStatistics { .empty() }
    public nonisolated var writeStatistics: OperationStatistics { .empty() }

    public func get(_ key: Key) -> Value? {
        storage[key]
    }

    public func put(_ key: Key, _ value: Value) {
        storage[key] = value
    }

    public func put(entries: [Key: Value]) {
        storage.merge(entries) { _, new in new }
    }

    public func getOrLoad(
        _ key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?
    ) async throws -> Value? {
        if let cached = storage[key] { return cached }
        guard let loaded = try await load(key) else { return nil }
        storage[key] = loaded
        return loaded
    }

    public func remove(_ key: Key) {
        storage.removeValue(forKey: key)
    }

    public func all() -> [Key: Value] {
        storage
    }

    public func clear() {
        storage.removeAll()
    }
}

/// File-based cache.
///
/// - Parameters:
///   - directory: directory for storing state files
///   - serdes: key and value de/serializers
public actor FileCache<Key: Hashable & Sendable, Value: Sendable>: Cache {
    private let directory: URL
    private let serdes: any CacheSerdes<Key, Value>
    private var inMemory: [Key: Value] = [:]
    private var readStats: OperationStatistics = .empty()
    private var writeStats: OperationStatistics = .empty()

    public init(directory: URL, serdes: any CacheSerdes<Key, Value>) {
        self.directory = directory
        self.serdes = serdes
    }

    public var readStatistics: OperationStatistics { readStats }
    public var writeStatistics: OperationStatistics { writeStats }

    public func get(_ key: Key) -> Value? {
        if let cached = inMemory[key] { return cached }
        let url = fileURL(for: key)
        let start = DispatchTime.now()
        guard let data = try? Data(contentsOf: url) else { return nil }
        let duration = Self.millis(since: start)
        readStats = readStats.updateWith(amount: data.count, duration: duration)
        guard let decoded = try? serdes.decodeValue(data) else { return nil }
        inMemory[key] = decoded
        return decoded
    }

    public func put(_ key: Key, _ value: Value) throws {
        inMemory[key] = value
        let start = DispatchTime.now()
        let data = try serdes.encodeValue(value)
        ensureDirectory()
        try data.write(to: fileURL(for: key), options: .atomic)
        let duration = Self.millis(since: start)
        writeStats = writeStats.updateWith(amount: data.count, duration: duration)
    }

    public func put(entries: [Key: Value]) throws {
        for (key, value) in entries {
            try put(key, value)
        }
    }

    public func getOrLoad(
        _ key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?
    ) async throws -> Value? {
        if let cached = get(key) { return cached }
        guard let loaded = try await load(key) else { return nil }
        try put(key, loaded)
        return loaded
    }

    public func remove(_ key: Key) throws {
        inMemory.removeValue(forKey: key)
        let url = fileURL(for: key)
        if FileManager.default.fileExists(atPath: url.path) {
            try FileManager.default.removeItem(at: url)
        }
    }

    public func all() -> [Key: Value] {
        var result: [Key: Value] = [:]
        for key in listKeysOnDisk() {
            if let value = get(key) {
                result[key] = value
            }
        }
        return result
    }

    public func clear() throws {
        inMemory.removeAll()
        for key in listKeysOnDisk() {
            try remove(key)
        }
    }

    private static func millis(since start: DispatchTime) -> Int64 {
        Int64((DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000)
    }

    private func ensureDirectory() {
        try? FileManager.default.createDirectory(
            at: directory,
            withIntermediateDirectories: true
        )
    }

    private func fileURL(for key: Key) -> URL {
        directory.appendingPathComponent(serdes.encodeKey(key))
    }

    private func listKeysOnDisk() -> [Key] {
        guard let urls = try? FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: nil
        ) else { return [] }
        return urls.compactMap { serdes.decodeKey($0.lastPathComponent) }
    }
}

/// Time-based refreshing cache.
///
/// After `interval` has passed, values are reloaded using the `load` closure
/// originally associated with the key on the first `getOrLoad(_:load:)` call.
/// On failure, the next refresh attempt uses a reduced interval up to
/// `maxRefreshIntervalOnFailure`.
public actor RefreshingCache<Key: Hashable & Sendable, Value: Sendable>: Cache {
    public static var refreshIntervalOnFailureReduction: Int { 10 }
    public static var maxRefreshIntervalOnFailure: TimeInterval { 5 }

    public typealias RefreshListener = @Sendable (Key, Value?) async -> Void

    private let underlying: any Cache<Key, Value>
    private let interval: TimeInterval
    private var jobs: [Key: Task<Void, Never>] = [:]
    private var listeners: [UUID: RefreshListener] = [:]

    public init(underlying: any Cache<Key, Value>, interval: TimeInterval) {
        self.underlying = underlying
        self.interval = interval
    }

    public var readStatistics: OperationStatistics {
        get async { await underlying.readStatistics }
    }

    public var writeStatistics: OperationStatistics {
        get async { await underlying.writeStatistics }
    }

    public func get(_ key: Key) async -> Value? {
        await underlying.get(key)
    }

    public func put(_ key: Key, _ value: Value) async throws {
        try await underlying.put(key, value)
    }

    public func put(entries: [Key: Value]) async throws {
        try await underlying.put(entries: entries)
    }

    public func getOrLoad(
        _ key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?
    ) async throws -> Value? {
        let value = try await underlying.getOrLoad(key, load: load)
        if value != nil, jobs[key] == nil {
            scheduleRefresh(key: key, load: load, after: interval)
        }
        return value
    }

    public func remove(_ key: Key) async throws {
        jobs.removeValue(forKey: key)?.cancel()
        try await underlying.remove(key)
    }

    public func all() async -> [Key: Value] {
        await underlying.all()
    }

    public func clear() async throws {
        for job in jobs.values { job.cancel() }
        jobs.removeAll()
        try await underlying.clear()
    }

    @discardableResult
    public func register(onEntryRefreshed listener: @escaping RefreshListener) -> UUID {
        let id = UUID()
        listeners[id] = listener
        return id
    }

    public func unregister(_ id: UUID) {
        listeners.removeValue(forKey: id)
    }

    private func scheduleRefresh(
        key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?,
        after delay: TimeInterval
    ) {
        let task = Task<Void, Never> { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            guard !Task.isCancelled else { return }
            await self?.performRefresh(key: key, load: load)
        }
        jobs[key] = task
    }

    private func performRefresh(
        key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?
    ) async {
        let result: Result<Value?, any Error>
        do {
            result = .success(try await load(key))
        } catch {
            result = .failure(error)
        }
        guard !Task.isCancelled else { return }

        switch result {
        case .success(let refreshed):
            if let refreshed {
                try? await underlying.put(key, refreshed)
            } else {
                try? await underlying.remove(key)
            }
            await notify(key: key, value: refreshed)
            rescheduleIfActive(key: key, load: load, after: interval)
        case .failure:
            let nextDelay = min(
                interval / Double(Self.refreshIntervalOnFailureReduction),
                Self.maxRefreshIntervalOnFailure
            )
            rescheduleIfActive(key: key, load: load, after: nextDelay)
        }
    }

    private func rescheduleIfActive(
        key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?,
        after delay: TimeInterval
    ) {
        guard !Task.isCancelled else { return }
        scheduleRefresh(key: key, load: load, after: delay)
    }

    private func notify(key: Key, value: Value?) async {
        let snapshot = Array(listeners.values)
        for listener in snapshot {
            await listener(key, value)
        }
    }
}

/// Statistics-tracking cache.
///
/// Statistics about cache hits and misses are kept and provided.
///
/// - Parameter underlying: cache that handles the actual data storage
public actor TrackingCache<Key: Hashable & Sendable, Value: Sendable>: Cache {
    private let underlying: any Cache<Key, Value>
    private var hitCount: Int64 = 0
    private var missCount: Int64 = 0

    public init(underlying: any Cache<Key, Value>) {
        self.underlying = underlying
    }

    /// Provides the current number of cache hits.
    public var hits: Int64 { hitCount }

    /// Provides the current number of cache misses.
    public var misses: Int64 { missCount }

    public var readStatistics: OperationStatistics {
        get async { await underlying.readStatistics }
    }

    public var writeStatistics: OperationStatistics {
        get async { await underlying.writeStatistics }
    }

    public func get(_ key: Key) async -> Value? {
        let result = await underlying.get(key)
        if result == nil {
            missCount += 1
        } else {
            hitCount += 1
        }
        return result
    }

    public func put(_ key: Key, _ value: Value) async throws {
        try await underlying.put(key, value)
    }

    public func put(entries: [Key: Value]) async throws {
        try await underlying.put(entries: entries)
    }

    public func getOrLoad(
        _ key: Key,
        load: @Sendable @escaping (Key) async throws -> Value?
    ) async throws -> Value? {
        if let cached = await get(key) { return cached }
        guard let loaded = try await load(key) else { return nil }
        try await put(key, loaded)
        return loaded
    }

    public func remove(_ key: Key) async throws {
        try await underlying.remove(key)
    }

    public func all() async -> [Key: Value] {
        await underlying.all()
    }

    public func clear() async throws {
        try await underlying.clear()
    }
}
